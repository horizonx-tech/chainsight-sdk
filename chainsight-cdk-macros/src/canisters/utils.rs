use std::path::PathBuf;

pub fn extract_contract_name_from_path(s: &str) -> String {
    let path = PathBuf::from(s);
    let name = path.file_stem().expect("file_stem failed");
    name.to_str().expect("to_str failed").to_string()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_extract_contract_name_from_path() {
        let path = "__interfaces/Oracle.json";
        assert_eq!(extract_contract_name_from_path(path), "Oracle");
    }
}
