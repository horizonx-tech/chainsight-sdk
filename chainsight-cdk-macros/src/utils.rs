use proc_macro::TokenStream;
use quote::quote;
use syn::parse_macro_input;

pub fn chainsight_common(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitInt);
    let output = quote! {
        #[derive(Clone, Debug, PartialEq, candid::CandidType, candid::Deserialize, serde::Serialize)]
        pub struct CanisterMetricsSnapshot {
            pub timestamp: u64,
            pub cycles: u128,
        }
        chainsight_cdk_macros::manage_vec_state!("canister_metrics_snapshot", CanisterMetricsSnapshot, false);

        fn setup_monitoring_canister_metrics() {
            let round_timestamp = |ts: u32, unit: u32| ts / unit * unit;
            let current_time_sec = (ic_cdk::api::time() / (1000 * 1000000)) as u32;
            let delay = round_timestamp(current_time_sec, #item) + #item - current_time_sec;

            ic_cdk_timers::set_timer(std::time::Duration::from_secs(delay as u64), move || {
                ic_cdk_timers::set_timer_interval(std::time::Duration::from_secs(#item as u64), || {
                    monitor_canister_metrics();
                });
            });
            // in first time, immediate execution
            monitor_canister_metrics();
        }

        fn monitor_canister_metrics() {
            let timestamp = ic_cdk::api::time();
            let cycles = ic_cdk::api::canister_balance128();
            let datum = CanisterMetricsSnapshot {
                timestamp,
                cycles,
            };
            ic_cdk::println!("monitoring: {:?}", datum.clone());
            add_canister_metrics_snapshot(datum);
        }

        #[ic_cdk_macros::init]
        fn init() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk_macros::post_upgrade]
        fn post_upgrade() {
            setup_monitoring_canister_metrics()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metric() -> CanisterMetricsSnapshot {
            get_last_canister_metrics_snapshot()
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        pub fn metrics(n: usize) -> Vec<CanisterMetricsSnapshot> {
            get_top_canister_metrics_snapshots(n)
        }

        async fn _get_target_proxy(target: candid::Principal) -> candid::Principal {
            let out: ic_cdk::api::call::CallResult<(candid::Principal,)> = ic_cdk::api::call::call(target, "get_proxy", ()).await;
            out.unwrap().0
        }
    };
    output.into()
}

pub fn did_export(input: TokenStream) -> TokenStream {
    let item = parse_macro_input!(input as syn::LitStr);
    let file_name = item.value() + ".did";
    TokenStream::from(quote! {
        candid::export_service!();
        #[ic_cdk::query(name = "__get_candid_interface_tmp_hack")]
        #[candid::candid_method(query, rename = "__get_candid_interface_tmp_hack")]
        fn __export_did_tmp_() -> String {
            __export_service()
        }
        #[cfg(test)]
        mod tests {
            use super::*;

            #[test]
            fn gen_candid() {
                std::fs::write(#file_name, __export_service()).unwrap();
            }
        }
    })
}
