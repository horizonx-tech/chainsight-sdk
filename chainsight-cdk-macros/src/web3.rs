use proc_macro::TokenStream;
use quote::quote;

pub fn define_transform_for_web3() -> TokenStream {
    let output = quote! {
        #[ic_cdk::query]
        #[candid::candid_method(query)]
        fn transform(response: ic_cdk::api::management_canister::http_request::TransformArgs) -> ic_cdk::api::management_canister::http_request::HttpResponse {
            let res = response.response;
            ic_cdk::api::management_canister::http_request::HttpResponse {
                status: res.status,
                headers: Vec::default(),
                body: res.body,
            }
        }
    };
    output.into()
}

pub fn define_web3_ctx() -> TokenStream {
    let output = quote! {
        #[derive(Default, Clone, Debug, PartialEq, candid::CandidType, candid::Deserialize)]
        pub struct Web3CtxParam {
            pub url: String,
            pub from: Option<String>,
            pub chain_id: u64,
            pub key: chainsight_cdk::types::EcdsaKeyEnvs,
        }
        manage_single_state!("web3_ctx_param", Web3CtxParam, false);

        pub fn web3_ctx() -> Result<ic_solidity_bindgen::Web3Context, ic_web3::Error> {
            let param = get_web3_ctx_param();
            let from = match param.from {
                Some(from) => Address::from_str(&from).unwrap(),
                None => Address::from_low_u64_be(0),
            };
            ic_solidity_bindgen::Web3Context::new(
                &param.url,
                from,
                param.chain_id,
                param.key.to_key_name(),
            )
        }
    };
    output.into()
}

pub fn define_get_ethereum_address() -> TokenStream {
    let output = quote! {
        fn default_derivation_key() -> Vec<u8> {
            ic_cdk::id().as_slice().to_vec()
        }
        async fn public_key(key_name: String) -> Result<Vec<u8>, String> {
            ic_web3::ic::get_public_key(None, vec![default_derivation_key()], key_name).await
        }
        async fn ethereum_address(key_name: String) -> Result<Address, String> {
            let pub_key = public_key(key_name).await?;
            ic_web3::ic::pubkey_to_address(&pub_key)
        }

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn get_ethereum_address(key: chainsight_cdk::types::EcdsaKeyEnvs) -> String {
            match ethereum_address(key.to_key_name()).await {
                Ok(v) => format!("0x{}", hex::encode(v)),
                Err(msg) => msg,
            }
        }
    };
    output.into()
}