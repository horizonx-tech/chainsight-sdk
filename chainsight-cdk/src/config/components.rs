use std::collections::HashMap;

use anyhow::bail;
use lazy_static::lazy_static;
use regex::Regex;

use crate::indexer::IndexingConfig;

lazy_static! {
    static ref MAPPING_CANDID_TY: HashMap<&'static str, &'static str> = [
        ("text", "String"),
        // ("blob", "&[u8]"),
        ("nat", "u128"),
        ("int", "i128"),
        ("nat8", "u8"),
        ("nat16", "u16"),
        ("nat32", "u32"),
        ("nat64", "u64"),
        ("int8", "i8"),
        ("int16", "i16"),
        ("int32", "i32"),
        ("int64", "i64"),
        ("float32", "f32"),
        ("float64", "f64"),
        ("bool", "bool"),
        // ("null", "()"),
    ].iter().cloned().collect();

    static ref REGEX_CANDID_FUNC: Regex = Regex::new(r"(?P<identifier>\w+)\s*:\s*\((?P<params>.*?)\)\s*(->\s*\((?P<return>.*?)\))?").unwrap();
    static ref REGEX_VECTOR: Regex = Regex::new(r"vec\s(?P<type>\w+)").unwrap();
    static ref REGEX_TUPLE: Regex = Regex::new(r"record\s\{\s(?P<items>(\w+(;\s|))+)\s\}").unwrap();
    static ref REGEX_STRUCT: Regex = Regex::new(r"(?P<field>\w+)\s*:\s*(?P<type>\w+)").unwrap();

    static ref REGEX_MULTIPLE_RECORD_TYPE: Regex = Regex::new(r"record\s*\{").unwrap();
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct CommonConfig {
    pub monitor_duration: u32,
    pub canister_name: String,
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct AlgorithmIndexerInput {
    pub method_name: String,
    pub response_type: String,
    pub source_type: AlgorithmInputType,
}

#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum AlgorithmInputType {
    #[serde(rename = "event_indexer")]
    EventIndexer,
    #[serde(rename = "key_value")]
    KeyValue,
    #[serde(rename = "key_values")]
    KeyValues,
}

impl Default for AlgorithmIndexerInput {
    fn default() -> Self {
        Self {
            method_name: "get_list".to_string(),
            response_type: "String".to_string(),
            source_type: AlgorithmInputType::EventIndexer,
        }
    }
}

impl Default for CommonConfig {
    fn default() -> Self {
        Self {
            monitor_duration: 3600,
            canister_name: "example_canister".to_string(),
        }
    }
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmIndexerConfig {
    pub common: CommonConfig,
    pub indexing: IndexingConfig,
    pub input: AlgorithmIndexerInput,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmLensConfig {
    pub common: CommonConfig,
    pub target_count: usize,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerConfig {
    pub common: CommonConfig,
    pub def: EventIndexerEventDefinition,
}
#[derive(Default, Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerEventDefinition {
    pub identifier: String,
    pub abi_file_path: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct RelayerConfig {
    pub common: CommonConfig,
    pub destination: String,
    pub oracle_type: String,
    pub method_name: String,
    pub canister_method_value_type: CanisterMethodValueType,
    pub abi_file_path: String,
    pub lens_targets: Option<LensTargets>,
}
#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct LensTargets {
    pub identifiers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub enum CanisterMethodValueType {
    Scalar(String, bool),                // struct name, is_scalar
    Tuple(Vec<(String, bool)>),          // struct_name, is_scalar
    Struct(Vec<(String, String, bool)>), // temp: Only non-nested `record` are supported.
    Vector(String, bool),                // struct_name, is_scalar
}
#[derive(Debug, Clone, PartialEq)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    pub params: Vec<String>,
    pub return_value: CanisterMethodValueType,
}

impl CanisterMethodIdentifier {
    pub fn parse_from_str(s: &str) -> anyhow::Result<Self> {
        let captures = REGEX_CANDID_FUNC.captures(s).ok_or(anyhow::anyhow!(
            "method.identifier does not satisfy the supported expression: {}, supplied={}",
            REGEX_CANDID_FUNC.to_string(),
            s
        ))?;

        let identifier = captures.name("identifier").unwrap().as_str().to_string();

        let params_str = captures.name("params").unwrap().as_str();
        let params_result: Vec<String> = if params_str.is_empty() {
            vec![]
        } else {
            params_str
                .split(',')
                .map(|s| convert_type_from_candid_type(s.trim()).0)
                .collect()
        };
        let params = params_result;

        let return_value_str = captures.name("return").unwrap().as_str();
        let return_value = Self::parse_return_value(return_value_str)?;

        Ok(CanisterMethodIdentifier {
            identifier,
            params,
            return_value,
        })
    }

    fn parse_return_value(s: &str) -> anyhow::Result<CanisterMethodValueType> {
        let record_type_count = REGEX_MULTIPLE_RECORD_TYPE.find_iter(s).count();
        if record_type_count >= 2 {
            bail!(
                "Sorry, Currently nested `record` types are not supported. This will be supported in the future.\nTarget literal = {}",
                s
            ); // TODO: Support nested `record` types.
        }
        // vector
        if s.starts_with("vec") {
            let captures = REGEX_VECTOR.captures(s);
            if let Some(captures_value) = captures {
                let ty = captures_value.name("type").unwrap().as_str();
                let val = convert_type_from_candid_type(ty);
                return Ok(CanisterMethodValueType::Vector(val.0, val.1));
            }
            bail!("Invalid candid's result types:{}", s);
        }

        // Scalar
        if !s.starts_with("record") {
            let val = convert_type_from_candid_type(s);
            return Ok(CanisterMethodValueType::Scalar(val.0, val.1));
        }

        // Tuple
        let captures = REGEX_TUPLE.captures(s);
        if let Some(captures_value) = captures {
            let items = captures_value.name("items").unwrap().as_str();
            let tuple_result: Vec<(String, bool)> = items
                .split(';')
                .map(|s| convert_type_from_candid_type(s.trim()))
                .collect();
            let tuple = tuple_result;
            return Ok(CanisterMethodValueType::Tuple(tuple));
        }

        // Struct
        let items = REGEX_STRUCT.captures_iter(s);
        let mut struct_items = vec![];
        for cap in items {
            let field = cap.name("field").unwrap().as_str().to_string();
            let ty = convert_type_from_candid_type(cap.name("type").unwrap().as_str());
            struct_items.push((field, ty.0, ty.1));
        }
        if struct_items.is_empty() {
            bail!("Invalid candid's result types: {}", s);
        }
        Ok(CanisterMethodValueType::Struct(struct_items))
    }
}

fn convert_type_from_candid_type(s: &str) -> (String, bool) {
    // ref: https://internetcomputer.org/docs/current/references/candid-ref
    let ty_str = MAPPING_CANDID_TY.get(&s);
    if let Some(ty_str) = ty_str {
        return (ty_str.to_string(), true);
    }
    (s.to_string(), false)
}
