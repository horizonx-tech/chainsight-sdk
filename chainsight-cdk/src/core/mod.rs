mod env;
pub use core::*;
pub use env::*;
pub use time::*;
mod types;
pub use types::*;
