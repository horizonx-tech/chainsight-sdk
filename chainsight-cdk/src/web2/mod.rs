#[warn(clippy::module_inception)]
mod web2;
pub use web2::*;
mod processors;
pub use processors::*;
