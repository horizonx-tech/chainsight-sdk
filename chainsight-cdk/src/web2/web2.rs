use std::collections::HashMap;

use ic_cdk::api::management_canister::http_request::{
    self, http_request, CanisterHttpRequestArgument, HttpHeader,
};
use serde::de::DeserializeOwned;
pub struct Web2HttpsSnapshotIndexer {
    pub url: String,
}

pub struct HttpsSnapshotParam {
    pub queries: HashMap<String, String>,
    pub headers: HashMap<String, String>,
}

impl Web2HttpsSnapshotIndexer {
    pub fn new(url: String) -> Self {
        Self { url }
    }

    pub async fn get<T, V>(&self, param: HttpsSnapshotParam) -> anyhow::Result<V>
    where
        V: DeserializeOwned,
    {
        let headers: Vec<HttpHeader> = param
            .headers
            .iter()
            .map(|(k, v)| HttpHeader {
                name: k.to_string(),
                value: v.to_string(),
            })
            .collect();
        let result = http_request(CanisterHttpRequestArgument {
            url: build_url(self.url.clone().as_str(), param.queries),
            method: http_request::HttpMethod::GET,
            headers,
            max_response_bytes: None,
            transform: None,
            body: None,
        })
        .await
        .expect("http_request failed");
        let res: V = serde_json::from_slice(&result.0.body)?;
        Ok(res)
    }
}

fn build_url(url: &str, queries: HashMap<String, String>) -> String {
    let mut url = url.to_string();
    if !queries.is_empty() {
        url.push('?');
        for (k, v) in queries {
            url.push_str(&format!("{}={}&", k, v));
        }
        url.pop();
    }
    url
}
