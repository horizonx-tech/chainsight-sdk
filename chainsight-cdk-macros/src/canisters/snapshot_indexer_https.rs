use chainsight_cdk::config::components::SnapshotIndexerHTTPSConfig;
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::parse_macro_input;

use crate::canisters::utils::generate_queries_without_timestamp;

pub fn def_snapshot_indexer_https(input: TokenStream) -> TokenStream {
    let input_json_string: String = parse_macro_input!(input as syn::LitStr).value();
    let config: SnapshotIndexerHTTPSConfig =
        serde_json::from_str(&input_json_string).expect("Failed to parse input_json_string");
    snapshot_indexer_https(config).into()
}

fn snapshot_indexer_https(config: SnapshotIndexerHTTPSConfig) -> proc_macro2::TokenStream {
    let use_idents = generate_use_idents(&config.common.canister_name);
    let custom = custom_code(config);
    quote! {
        #use_idents
        #custom
    }
}

fn generate_use_idents(id: &str) -> proc_macro2::TokenStream {
    let id_ident = format_ident!("{}", id);

    quote! {
        use std::collections::HashMap;

        use chainsight_cdk::core::HttpsSnapshotIndexerSourceAttrs;
        use chainsight_cdk::web2::{HttpsSnapshotParam, Web2HttpsSnapshotIndexer};
        use chainsight_cdk_macros::{
            chainsight_common, did_export, init_in, prepare_stable_structure, stable_memory_for_vec,
            timer_task_func, snapshot_indexer_https_source, StableMemoryStorable,
        };
        use candid::{Decode, Encode};
        use #id_ident::*;
    }
}

fn custom_code(config: SnapshotIndexerHTTPSConfig) -> proc_macro2::TokenStream {
    let SnapshotIndexerHTTPSConfig {
        common,
        url,
        headers,
        queries,
    } = config;

    let id = &common.canister_name;
    let header_keys: Vec<String> = headers.keys().cloned().collect();
    let header_values: Vec<String> = headers.values().cloned().collect();
    let query_keys: Vec<String> = queries.keys().cloned().collect();
    let query_values: Vec<String> = queries.values().cloned().collect();
    let queries = generate_queries_without_timestamp(format_ident!("SnapshotValue"));

    quote! {
        did_export!(#id); // NOTE: need to be declared before query, update
        init_in!();
        chainsight_common!();
        snapshot_indexer_https_source!();

        #[derive(Debug, Clone, candid::CandidType, candid::Deserialize, serde::Serialize, StableMemoryStorable)]
        #[stable_mem_storable_opts(max_size = 10000, is_fixed_size = false)] // temp: max_size
        pub struct Snapshot {
            pub value: SnapshotValue,
            pub timestamp: u64,
        }
        prepare_stable_structure!();
        stable_memory_for_vec!("snapshot", Snapshot, 0, true);
        timer_task_func!("set_task", "index");

        const URL : &str = #url;
        fn get_attrs() -> HttpsSnapshotIndexerSourceAttrs {
            HttpsSnapshotIndexerSourceAttrs {
                queries: HashMap::from([
                    #(
                        (#query_keys.to_string(), #query_values.to_string()),
                    )*
                ]),
            }
        }

        #[ic_cdk::query]
        #[candid::candid_method(query)]
        fn transform_https_response(response: ic_cdk::api::management_canister::http_request::TransformArgs) -> ic_cdk::api::management_canister::http_request::HttpResponse {
            use chainsight_cdk::web3::TransformProcessor;
            let processor = chainsight_cdk::web2::processors::HTTPSResponseTransformProcessor::<SnapshotValue>::new();
            processor.transform(response)
        }

        #[ic_cdk::update]
        #[candid::candid_method(update)]
        async fn index() {
            if ic_cdk::caller() != proxy() {
                panic!("Not permitted")
            }
            let indexer = Web2HttpsSnapshotIndexer::new(
                URL.to_string(),
            );
            let res = indexer.get::<String, SnapshotValue>(
                HttpsSnapshotParam {
                    headers: vec![
                        #(
                            (#header_keys.to_string(), #header_values.to_string()),
                        )*
                    ].into_iter().collect(),
                    queries: vec![
                        #(
                            (#query_keys.to_string(), #query_values.to_string()),
                        )*
                    ].into_iter().collect(),
                }
            ).await.expect("Failed to get by indexer");
            let snapshot = Snapshot {
                value: res,
                timestamp: ic_cdk::api::time() / 1000000,
            };
            let _ = add_snapshot(snapshot.clone());
        }
        #queries
    }
}

#[cfg(test)]
mod test {
    use std::collections::BTreeMap;

    use chainsight_cdk::config::components::CommonConfig;
    use insta::assert_display_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot() {
        let config = SnapshotIndexerHTTPSConfig {
            common: CommonConfig {
                canister_name: "sample_snapshot_indexer_https".to_string(),
            },
            url: "https://api.coingecko.com/api/v3/simple/price".to_string(),
            headers: BTreeMap::from([("content-type".to_string(), "application/json".to_string())]),
            queries: BTreeMap::from([
                ("ids".to_string(), "dai".to_string()),
                ("vs_currencies".to_string(), "usd".to_string()),
            ]),
        };
        let generated = snapshot_indexer_https(config);
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_display_snapshot!("snapshot__snapshot_indexer_https", formatted);
    }
}
