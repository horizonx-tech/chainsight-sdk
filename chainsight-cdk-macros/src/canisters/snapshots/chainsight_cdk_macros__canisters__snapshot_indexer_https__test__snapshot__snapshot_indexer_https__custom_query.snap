---
source: chainsight-cdk-macros/src/canisters/snapshot_indexer_https.rs
expression: formatted
---
use candid::{Decode, Encode};
use chainsight_cdk::core::HttpsSnapshotIndexerSourceAttrs;
use chainsight_cdk::web2::{HttpsSnapshotParam, Web2HttpsSnapshotIndexer};
use chainsight_cdk_macros::{
    chainsight_common, did_export, init_in, manage_single_state, prepare_stable_structure,
    snapshot_indexer_https_source, stable_memory_for_btree_map, stable_memory_for_scalar,
    timer_task_func, CborSerde, StableMemoryStorable,
};
use ic_stable_structures::writer::Writer;
use sample_snapshot_indexer_https::*;
use std::collections::HashMap;
did_export!("sample_snapshot_indexer_https");
init_in!(2);
chainsight_common!();
snapshot_indexer_https_source!();
#[derive(
    Debug,
    Clone,
    candid :: CandidType,
    candid :: Deserialize,
    serde :: Serialize,
    StableMemoryStorable,
)]
#[stable_mem_storable_opts(max_size = 10000, is_fixed_size = false)]
pub struct Snapshot {
    pub value: SnapshotValue,
    pub timestamp: u64,
}
prepare_stable_structure!();
stable_memory_for_btree_map!("snapshot", Snapshot, 1, true);
timer_task_func!("set_task", "index", 3);
const URL: &str = "https://api.coingecko.com/api/v3/simple/price";
fn get_attrs() -> HttpsSnapshotIndexerSourceAttrs {
    HttpsSnapshotIndexerSourceAttrs {
        queries: get_queries()
            .into_iter()
            .collect::<HashMap<String, String>>(),
    }
}
#[ic_cdk::query]
#[candid::candid_method(query)]
fn transform_https_response(
    response: ic_cdk::api::management_canister::http_request::TransformArgs,
) -> ic_cdk::api::management_canister::http_request::HttpResponse {
    use chainsight_cdk::web3::TransformProcessor;
    let processor =
        chainsight_cdk::web2::processors::HTTPSResponseTransformProcessor::<SnapshotValue>::new();
    processor.transform(response)
}
#[ic_cdk::update]
#[candid::candid_method(update)]
async fn index() {
    if ic_cdk::caller() != proxy() {
        panic!("Not permitted")
    }
    let indexer = Web2HttpsSnapshotIndexer::new(URL.to_string());
    let res = indexer
        .get::<String, SnapshotValue>(HttpsSnapshotParam {
            headers: vec![("content-type".to_string(), "application/json".to_string())]
                .into_iter()
                .collect(),
            queries: get_queries()
                .into_iter()
                .collect::<HashMap<String, String>>(),
        })
        .await
        .expect("Failed to get by indexer");
    let snapshot = Snapshot {
        value: res,
        timestamp: ic_cdk::api::time() / 1000000,
    };
    add_snapshot(snapshot.clone());
    ic_cdk::println!(
        "timestamp={}, value={:?}",
        snapshot.timestamp,
        snapshot.value
    );
}
fn _get_last_snapshot_value() -> SnapshotValue {
    get_last_snapshot().value
}
fn _get_top_snapshot_values(n: u64) -> Vec<SnapshotValue> {
    get_top_snapshots(n)
        .iter()
        .map(|s| s.value.clone())
        .collect()
}
fn _get_snapshot_value(idx: u64) -> SnapshotValue {
    get_snapshot(idx).value
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_last_snapshot_value() -> SnapshotValue {
    _get_last_snapshot_value()
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_top_snapshot_values(n: u64) -> Vec<SnapshotValue> {
    _get_top_snapshot_values(n)
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_snapshot_value(idx: u64) -> SnapshotValue {
    _get_snapshot_value(idx)
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_last_snapshot_value(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProviderWithoutArgs::<SnapshotValue>::new(
        proxy(),
        _get_last_snapshot_value,
    )
    .reply(input)
    .await
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_top_snapshot_values(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProvider::<u64, Vec<SnapshotValue>>::new(
        proxy(),
        _get_top_snapshot_values,
    )
    .reply(input)
    .await
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_snapshot_value(input: std::vec::Vec<u8>) -> std::vec::Vec<u8> {
    use chainsight_cdk::rpc::Receiver;
    chainsight_cdk::rpc::ReceiverProvider::<u64, SnapshotValue>::new(proxy(), _get_snapshot_value)
        .reply(input)
        .await
}

